src/analyses/cfg_dominators.h

```diff
@@ -158,12 +183,39 @@ void cfg_dominators_templatet<P, T, post_dom>::fixedpoint(P &program)
     {
       for(const auto &edge : (post_dom ? node.in : node.out))
       {
-        worklist.push_back(cfg[edge.first].PC);
+        worklist.push_back(edge.first);
       }
     }
   }
 }
 
+template <class P, class T, bool post_dom>
+bool cfg_dominators_templatet<P, T, post_dom>::dominates(T lhs, T rhs) const
+{
+  if(lhs == rhs)
+    return true;
+
+  const auto lhs_block = cfg.entry_map.at(lhs);
+  const auto rhs_block = cfg.entry_map.at(rhs);
+
+  if(lhs_block != rhs_block)
+    return basic_block_dominates(lhs_block, rhs_block);
+
+  // Special case when the program points belong to the same block: lhs
+  // dominates rhs iff it is <= rhs in program order (or the reverse if we're
+  // a postdominator analysis)
+
+  for(const auto &instruction : cfg[lhs_block].block)
+  {
+    if(instruction == lhs)
+      return !post_dom;
+    else if(instruction == rhs)
+      return post_dom;
+  }
+
+  UNREACHABLE; // Entry map is inconsistent with block members?
```

Comment:
:pick: I think in this case it might be better to just use a regular invariant and put the comment on that instead